rott/vgatext.c

```diff
@@ -0,0 +1,198 @@
+/*
+MIT License
+
+Copyright (c) 2024 erysdren (it/she/they)
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of
+this software and associated documentation files (the "Software"), to deal in
+the Software without restriction, including without limitation the rights to
+use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
+of the Software, and to permit persons to whom the Software is furnished to do
+so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+#include <SDL2/SDL.h>
+
+#include "vga8x16.h"
+
+#define BLINK_HZ ((1000 / 70) * 16)
+
+static const Uint8 palette[16][3] = {
+	{0x00, 0x00, 0x00}, {0x00, 0x00, 0xab}, {0x00, 0xab, 0x00}, {0x00, 0xab, 0xab},
+	{0xab, 0x00, 0x00}, {0xab, 0x00, 0xab}, {0xab, 0x57, 0x00}, {0xab, 0xab, 0xab},
+	{0x57, 0x57, 0x57}, {0x57, 0x57, 0xff}, {0x57, 0xff, 0x57}, {0x57, 0xff, 0xff},
+	{0xff, 0x57, 0x57}, {0xff, 0x57, 0xff}, {0xff, 0xff, 0x57}, {0xff, 0xff, 0xff}
+};
+
+static void render_cell(Uint8 *image, int pitch, Uint16 cell, SDL_bool noblink)
+{
+	// get components
+	Uint8 code = (Uint8)(cell & 0xFF);
+	Uint8 attr = (Uint8)(cell >> 8);
+
+	// break out attributes
+	Uint8 blink = (attr >> 7) & 0x01;
+	Uint8 bgcolor = (attr >> 4) & 0x07;
+	Uint8 fgcolor = attr & 0x0F;
+
+	for (int y = 0; y < 16; y++)
+	{
+		for (int x = 0; x < 8; x++)
+		{
+			// write bgcolor
+			image[y * pitch + x] = bgcolor;
+
+			if (blink && noblink)
+				continue;
+
+			// write fgcolor
+			Uint8 *bitmap = &VGA8X16[code * 16];
+
+			if (bitmap[y] & 1 << abs(x - 7))
+				image[y * pitch + x] = fgcolor;
+		}
+	}
+}
+
+int vgatext_main(SDL_Window *window, Uint16 *screen)
+{
+	Uint8 image0[400][640];
+	Uint8 image1[400][640];
+	Uint32 format, rmask, gmask, bmask, amask;
+	int bpp;
+	SDL_Renderer *renderer;
+	SDL_Texture *texture;
+	SDL_Surface *surface8;
+	SDL_Surface **windowsurface;
+	SDL_Surface *windowsurface0;
+	SDL_Surface *windowsurface1;
+	SDL_Color colors[16];
+	SDL_Rect rect;
+
+	if (!window || !screen)
+		return -1;
+
+	// setup renderer
+	renderer = SDL_GetRenderer(window);
+	if (!renderer)
+		return -1;
+
+	SDL_RenderSetLogicalSize(renderer, 640, 400);
+	SDL_SetWindowMinimumSize(window, 640, 400);
+	SDL_SetRenderDrawColor(renderer, 0, 0, 0, 255);
+	SDL_RenderClear(renderer);
+	SDL_RenderPresent(renderer);
+	SDL_SetHint(SDL_HINT_RENDER_SCALE_QUALITY, "nearest");
```

Comment:
This is really *very* blocky. How about "linear" instead?

Reply:
Sure!